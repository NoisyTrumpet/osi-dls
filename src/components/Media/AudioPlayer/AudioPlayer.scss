.audio-player {
  height: $space-48;
  line-height: $space-48;

  .audio-player__button-container {
    height: $space-48;
    width: $space-48;
    padding: $space-4;
    float: left;

    svg {
      position: absolute;
      overflow: visible;

      circle {
        transform: rotate(-90deg);
        transform-origin: 50% 50%;
        stroke-linecap: round;
        stroke: map-get($colors, 'primary');
        fill: transparent;
      }
    }

    .btn-icon {
      position: absolute;
      margin: $space-4;
    }
  }

  .typography {
    margin-left: $space-8;
    line-height: $space-24;
  }
}
