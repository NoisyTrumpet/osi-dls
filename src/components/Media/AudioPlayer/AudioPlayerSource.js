import React from 'react';
import PropTypes from 'prop-types';

const AudioPlayerSource = ({ src, type }) => (<source src={src} type={type} />);

AudioPlayerSource.defaultProps = {
  type: 'audio/mpeg',
};

AudioPlayerSource.propTypes = {
  src: PropTypes.string,
  type: PropTypes.oneOf([
    'audio/mpeg',
    'audio/ogg',
    'audio/wav',
  ]),
};

export default AudioPlayerSource;
