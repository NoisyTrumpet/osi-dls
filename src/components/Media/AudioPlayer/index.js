export { default as AudioPlayer } from './AudioPlayer'; // eslint-disable-line import/prefer-default-export
