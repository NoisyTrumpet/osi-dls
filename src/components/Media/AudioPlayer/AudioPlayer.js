/* eslint jsx-a11y/media-has-caption: 0 */

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { IconButton } from '../../IconButton';
import { Typography } from '../../Typography';
import AudioPlayerSource from './AudioPlayerSource';
import './AudioPlayer.scss';

const AudioPlayer = ({ addClass, children }) => {
  const audioRef = React.useRef(null);
  const [playing, setPlaying] = React.useState(false);
  const [circleProgress, setCircleProgress] = React.useState(0);
  const [progressAnimation, setProgressAnimation] = React.useState(null);

  const classes = classNames(
    'audio-player',
    addClass,
  );

  // Audio Actions
  const togglePlayback = () => { setPlaying(!playing); };
  const playAudio = () => { audioRef.current.play(); };
  const pauseAudio = () => { audioRef.current.pause(); };

  // Audio Event Handlers
  const handleEndedAudio = React.useCallback(() => {
    setPlaying(false);
    setCircleProgress(0);
  }, []);

  React.useEffect(() => {
    audioRef.current.addEventListener('ended', handleEndedAudio);
    return () => {
      audioRef.current.removeEventListener('ended', handleEndedAudio);
    };
  }, [handleEndedAudio]);

  // Animations
  const performAnimation = () => {
    if (audioRef.current && audioRef.current.duration) {
      setProgressAnimation(requestAnimationFrame(performAnimation));
      const percent = audioRef.current.currentTime / audioRef.current.duration;
      setCircleProgress(percent);
    }
  };
  const startAnimation = () => { requestAnimationFrame(performAnimation); };
  const stopAnimation = () => { cancelAnimationFrame(progressAnimation); };

  // 'playing' React hook
  React.useEffect(() => {
    if (playing) {
      playAudio();
      startAnimation();
    } else {
      pauseAudio();
      stopAnimation();
    }
  }, [playing]);

  // AudioPlayer Progress Circle
  const radius = 20;
  const stroke = 2;
  const circumference = radius * 2 * Math.PI;
  const strokeDashoffset = circumference - circleProgress * circumference;
  const ProgressCircle = (
    <svg
      alt=""
      role="img"
      focusable="false"
      aria-hidden="true"
      height={radius * 2}
      width={radius * 2}
    >
      <title>Audio Progress Circle</title>
      <circle
        strokeWidth={stroke}
        strokeDasharray={`${circumference} ${circumference}`}
        style={{ strokeDashoffset }}
        r={radius}
        cx={radius}
        cy={radius}
      />
    </svg>
  );

  // Typography and Icon Button Screen reader text
  const buttonActionText = `${playing ? 'Pause' : 'Play'} Audio`;
  const screenReaderText = `${playing ? 'Pause Audio' : `${Math.floor(circleProgress * 100)}% played, press play to continue`}`;

  return (
    <div className={classes}>
      <audio ref={audioRef}>{children}</audio>
      <div className="audio-player__button-container">
        {ProgressCircle}
        <IconButton
          icon={playing ? 'pausesmall-bold' : 'playsmall-bold'}
          variant="tiny"
          click={togglePlayback}
          srtAriaLive="polite"
        >
          {screenReaderText}
        </IconButton>
      </div>
      <Typography variant="body-medium">{buttonActionText}</Typography>
    </div>
  );
};

AudioPlayer.Source = AudioPlayerSource;

AudioPlayer.defaultProps = {};

AudioPlayer.propTypes = {
  children: (props, propName, componentName) => {
    let error;
    const prop = props[propName];
    const requiredChildType = 'AudioPlayerSource';
    React.Children.forEach(prop, (child) => {
      // type.name seems to work for both Class and Functional components
      if (child.type.name !== requiredChildType) {
        error = new Error(`\`${componentName}\` only accepts children of type \`${requiredChildType}\`. Child of type ${child.type.name} was found.`);
      }
    });
    return error;
  },
  addClass: PropTypes.string,
};

export default AudioPlayer;
