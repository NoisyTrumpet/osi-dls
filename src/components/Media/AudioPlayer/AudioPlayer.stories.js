import React from "react";
import { storiesOf } from "@storybook/react";
import { withSmartKnobs } from "storybook-addon-smart-knobs";
import { AudioPlayer } from "../AudioPlayer";

const parameters = {
  component: AudioPlayer,
  componentSubtitle: "The <AudioPlayer /> tag defines an Audio Player.",
  jest: ["AudioPlayer"]
};

storiesOf("components | Media/AudioPlayer", module)
  .addParameters(parameters)
  .addDecorator(withSmartKnobs({ ignoreProps: ["children"] }))
  .add("MP3 Audio", () => (
    <>
      {/* http://soundbible.com/2030-Daydreaming.html */}
      <AudioPlayer>
        <AudioPlayer.Source src="http://soundbible.com/mp3/Tinkle-Lisa_Redfern-1916445296.mp3" />
      </AudioPlayer>
    </>
  ))
  .add("WAV Audio", () => (
    <>
      {/* http://soundbible.com/1782-Tyrannosaurus-Rex-.html */}
      <AudioPlayer>
        <AudioPlayer.Source
          src="http://soundbible.com/wav/Tyrannosaurus%20Rex%20-SoundBible.com-45786848.wav"
          type="audio/wav"
        />
      </AudioPlayer>
    </>
  ));
