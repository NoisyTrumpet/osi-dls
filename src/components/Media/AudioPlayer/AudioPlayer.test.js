import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AudioPlayer } from '../AudioPlayer';

describe('<AudioPlayer />', () => {
  test('should render', () => {
    const wrapper = shallow(
      <AudioPlayer>
        <AudioPlayer.Source src="http://soundbible.com/mp3/Tinkle-Lisa_Redfern-1916445296.mp3" />
      </AudioPlayer>,
    );
    expect(wrapper.getElement()).toBeTruthy();
    expect(wrapper.find('audio').length).toBe(1);
    expect(wrapper.find('AudioPlayerSource').length).toBe(1);
    expect(wrapper.find('svg').length).toBe(1);
    expect(wrapper.find('circle').length).toBe(1);
    console.log(wrapper.debug());
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render mulitple sources', () => {
    const wrapper = shallow(
      <AudioPlayer>
        <AudioPlayer.Source src="http://soundbible.com/mp3/Tinkle-Lisa_Redfern-1916445296.mp3" />
        <AudioPlayer.Source
          src="http://soundbible.com/wav/Tinkle-Lisa_Redfern-1916445296.wav"
          type="audio/wav"
        />
      </AudioPlayer>,
    );
    expect(wrapper.getElement()).toBeTruthy();
    expect(wrapper.find('audio').length).toBe(1);
    expect(wrapper.find('AudioPlayerSource').length).toBe(2);
    expect(wrapper.find('svg').length).toBe(1);
    expect(wrapper.find('circle').length).toBe(1);
    console.log(wrapper.debug());
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
