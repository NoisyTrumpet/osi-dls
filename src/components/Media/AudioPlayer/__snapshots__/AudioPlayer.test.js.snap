// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer /> should render 1`] = `
<div
  className="audio-player"
>
  <audio>
    <AudioPlayerSource
      src="http://soundbible.com/mp3/Tinkle-Lisa_Redfern-1916445296.mp3"
      type="audio/mpeg"
    />
  </audio>
  <div
    className="audio-player__button-container"
  >
    <svg
      alt=""
      aria-hidden="true"
      focusable="false"
      height={40}
      role="img"
      width={40}
    >
      <title>
        Audio Progress Circle
      </title>
      <circle
        cx={20}
        cy={20}
        r={20}
        strokeDasharray="125.66370614359172 125.66370614359172"
        strokeWidth={2}
        style={
          Object {
            "strokeDashoffset": 125.66370614359172,
          }
        }
      />
    </svg>
    <ForwardRef
      click={[Function]}
      disabled={false}
      icon="playsmall-bold"
      size={16}
      srtAriaLive="polite"
      variant="tiny"
    >
      0% played, press play to continue
    </ForwardRef>
  </div>
  <Typography
    variant="body-medium"
  >
    Play Audio
  </Typography>
</div>
`;

exports[`<AudioPlayer /> should render mulitple sources 1`] = `
<div
  className="audio-player"
>
  <audio>
    <AudioPlayerSource
      src="http://soundbible.com/mp3/Tinkle-Lisa_Redfern-1916445296.mp3"
      type="audio/mpeg"
    />
    <AudioPlayerSource
      src="http://soundbible.com/wav/Tinkle-Lisa_Redfern-1916445296.wav"
      type="audio/wav"
    />
  </audio>
  <div
    className="audio-player__button-container"
  >
    <svg
      alt=""
      aria-hidden="true"
      focusable="false"
      height={40}
      role="img"
      width={40}
    >
      <title>
        Audio Progress Circle
      </title>
      <circle
        cx={20}
        cy={20}
        r={20}
        strokeDasharray="125.66370614359172 125.66370614359172"
        strokeWidth={2}
        style={
          Object {
            "strokeDashoffset": 125.66370614359172,
          }
        }
      />
    </svg>
    <ForwardRef
      click={[Function]}
      disabled={false}
      icon="playsmall-bold"
      size={16}
      srtAriaLive="polite"
      variant="tiny"
    >
      0% played, press play to continue
    </ForwardRef>
  </div>
  <Typography
    variant="body-medium"
  >
    Play Audio
  </Typography>
</div>
`;
